### PR TITLE
Add readability-implicit-bool-conversion

### PR DESCRIPTION
**Pull request by masterleinad**
It's rather annoying/confusing seeing `int`s used instead of bools.

### CODE REVIEW

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 1]
```diff
@@ -450,7 +450,7 @@ struct SubviewExtents {
   // Cannot declare zero-length arrays
   // '+' is used to silence GCC 7.2.0 -Wduplicated-branches warning when
   // RangeRank=1
-  enum { InternalRangeRank = RangeRank ? RangeRank : +1u };
+  enum { InternalRangeRank = RangeRank > 0 ? RangeRank : +1u };
```
Comment (PhilMiller) [inline core/src/impl/Kokkos_ViewMapping.hpp:453]: Did you want to change this to a `constexpr` as well?

File: core/src/impl/Kokkos_hwloc.cpp [hunk 2]
```diff
@@ -162,7 +162,7 @@ unsigned thread_mapping(const char* const label, const bool allow_async,
   const unsigned threads_per_core =
       thread_count / (use_numa_count * use_cores_per_numa);
 
-  if (thread_spawn_synchronous) {
+  if (thread_spawn_synchronous != 0) {
```
Comment (PhilMiller) [inline core/src/impl/Kokkos_hwloc.cpp:165]: Why not convert `thread_spawn_synchronous` to a `bool`? It's assigned 0 or 1 just above based on a ternary and some conditions.

**Review by PhilMiller — APPROVED** [review]

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Let's see what everyone else is thinking about this.

File: core/src/Cuda/Kokkos_Cuda_Instance.hpp [hunk 3]
```diff
@@ -109,7 +109,7 @@ class CudaInternal {
   mutable std::size_t m_scratchUnifiedCount;
   mutable std::size_t m_scratchFunctorSize;
 
-  size_type m_scratchUnifiedSupported;
+  bool m_scratchUnifiedSupported;
```
Comment (crtrott) [inline core/src/Cuda/Kokkos_Cuda_Instance.hpp:112]: I am unsure about this change due to alignment stuff. But I guess we don't generally avoid bool members.

File: core/src/impl/Kokkos_Core.cpp [hunk 4]
```diff
@@ -221,7 +221,7 @@ int get_ctest_gpu(const char* local_rank_str) {
   }
 
   auto const* comma = std::strchr(resource_str, ',');
-  if (!comma || strncmp(resource_str, "id:", 3)) {
+  if (!comma || strncmp(resource_str, "id:", 3) != 0) {
```
Comment (crtrott) [inline core/src/impl/Kokkos_Core.cpp:224]: maybe extra paren around `strncmp(...) != 0`

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 5]
```diff
@@ -2751,7 +2751,8 @@ struct ViewDataHandle<
   KOKKOS_INLINE_FUNCTION
   static handle_type assign(value_type* arg_data_ptr,
                             track_type const& /*arg_tracker*/) {
-    if (reinterpret_cast<uintptr_t>(arg_data_ptr) % Impl::MEMORY_ALIGNMENT) {
+    if (reinterpret_cast<uintptr_t>(arg_data_ptr) % Impl::MEMORY_ALIGNMENT >
```
Comment (crtrott) [inline core/src/impl/Kokkos_ViewMapping.hpp:2754]: extra paren?

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 6]
```diff
@@ -2762,7 +2763,8 @@ struct ViewDataHandle<
   KOKKOS_INLINE_FUNCTION
   static handle_type assign(handle_type const arg_data_ptr, size_t offset) {
     if (reinterpret_cast<uintptr_t>(arg_data_ptr + offset) %
-        Impl::MEMORY_ALIGNMENT) {
```
Comment (crtrott) [inline core/src/impl/Kokkos_ViewMapping.hpp:2765]: extra paren

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 7]
```diff
@@ -2796,7 +2798,8 @@ struct ViewDataHandle<
   KOKKOS_INLINE_FUNCTION
   static value_type* assign(value_type* arg_data_ptr,
                             track_type const& /*arg_tracker*/) {
-    if (reinterpret_cast<uintptr_t>(arg_data_ptr) % Impl::MEMORY_ALIGNMENT) {
+    if (reinterpret_cast<uintptr_t>(arg_data_ptr) % Impl::MEMORY_ALIGNMENT >
```
Comment (crtrott) [inline core/src/impl/Kokkos_ViewMapping.hpp:2801]: extra paren

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 8]
```diff
@@ -2807,7 +2810,8 @@ struct ViewDataHandle<
   KOKKOS_INLINE_FUNCTION
   static value_type* assign(handle_type const arg_data_ptr, size_t offset) {
     if (reinterpret_cast<uintptr_t>(arg_data_ptr + offset) %
-        Impl::MEMORY_ALIGNMENT) {
```
Comment (crtrott) [inline core/src/impl/Kokkos_ViewMapping.hpp:2810]: extra paren

File: core/src/impl/Kokkos_hwloc.cpp [hunk 9]
```diff
@@ -108,13 +108,13 @@ unsigned thread_mapping(const char* const label, const bool allow_async,
 
   const bool valid_numa = use_numa_count <= avail_numa_count;
   const bool valid_cores =
-      use_cores_per_numa && use_cores_per_numa <= avail_cores_per_numa;
+      use_cores_per_numa > 0 && use_cores_per_numa <= avail_cores_per_numa;
```
Comment (crtrott) [inline core/src/impl/Kokkos_hwloc.cpp:111]: extra paren

File: core/src/impl/Kokkos_hwloc.cpp [hunk 10]
```diff
@@ -108,13 +108,13 @@ unsigned thread_mapping(const char* const label, const bool allow_async,
...
   const bool valid_threads =
-      thread_count && thread_count <= use_numa_count * use_cores_per_numa *
-                                          avail_threads_per_core;
-  const bool balanced_numa = !(thread_count % use_numa_count);
+      thread_count > 0 && thread_count <= use_numa_count * use_cores_per_numa *
```
Comment (crtrott) [inline core/src/impl/Kokkos_hwloc.cpp:113]: extra paren

File: core/unit_test/TestHalfOperators.hpp [hunk 11]
```diff
@@ -882,14 +882,17 @@ struct Functor_TestHalfOperators {
     }
 
     // TODO: figure out why operator{!,&&,||} are returning __nv_bool
-    actual_lhs(NEG)   = static_cast<double>(!h_lhs);
-    expected_lhs(NEG) = !d_lhs;
+    actual_lhs(NEG)   = static_cast<double>(!static_cast<bool>(h_lhs));
```
Comment (crtrott) [inline core/unit_test/TestHalfOperators.hpp:885]: this is wrong: we actually WANT to test implicit conversion for various usage scenarios

File: core/unit_test/TestHalfOperators.hpp [hunk 12]
```diff
@@ -882,14 +882,17 @@ struct Functor_TestHalfOperators {
...
+    expected_lhs(NEG) = !static_cast<bool>(d_lhs);
 
-    actual_lhs(AND)   = static_cast<double>(half_type(0) && h_lhs);
-    expected_lhs(AND) = double(0) && d_lhs;
+    actual_lhs(AND) = static_cast<double>(static_cast<bool>(half_t(0)) &&
```
Comment (crtrott) [inline core/unit_test/TestHalfOperators.hpp:888]: revert this: purpose is to check implicit conversion to bool for use with && operator

File: core/unit_test/TestHalfOperators.hpp [hunk 13]
```diff
@@ -882,14 +882,17 @@ struct Functor_TestHalfOperators {
...
+                                          static_cast<bool>(h_lhs));
+    expected_lhs(AND) =
+        static_cast<bool>(double(0)) && static_cast<bool>(d_lhs);
 
-    actual_lhs(OR)   = static_cast<double>(h_lhs || half_type(1));
```
Comment (crtrott) [inline core/unit_test/TestHalfOperators.hpp:891]: revert: this is meant to test implicit conversion in use case of ||

**Review by crtrott — CHANGES_REQUESTED** [review]
Couple of changes I would want: 
1) some places now could use extra parenthesis for readability. 
2) The half test changes need to reverted since they test operators. 

Not sure about the enforcement.

File: core/src/OpenMP/Kokkos_OpenMP_Exec.hpp [hunk 14]
```diff
@@ -133,7 +133,7 @@ inline bool OpenMP::impl_is_initialized() noexcept {
 
 inline bool OpenMP::in_parallel(OpenMP const&) noexcept {
   // t_openmp_instance is only non-null on a master thread
-  return !Impl::t_openmp_instance ||
+  return Impl::t_openmp_instance == nullptr ||
```
Comment (dalg24) [inline core/src/OpenMP/Kokkos_OpenMP_Exec.hpp:136]: Not a good change IMO

**Review by dalg24 — CHANGES_REQUESTED** [review]
I find `if (!p) {...}` more readable than `if (p != nullptr) {...}`
Set `AllowPointerConditions` to `true` and revert these changes

**Comment by masterleinad** [comment]
> Set AllowPointerConditions to true and revert these changes

I knew that you didn't like it so it was always true. I don't know why it complained about that one instance. Do you care enough that you want a `NOLINT` comment for it?

**Comment by masterleinad** [comment]
@crtrott I addressed your comments.

File: core/src/OpenMP/Kokkos_OpenMP_Exec.hpp (hunk 14 above)
Comment (masterleinad) [reply core/src/OpenMP/Kokkos_OpenMP_Exec.hpp]: I added a NOLINTNEXTLINE for it.

**Review by crtrott — APPROVED** [review]

**Review by dalg24 — COMMENTED** [review]
IMO, all the casting to bool for things like `enum : bool { accessible = ... };` are bad changes.  They can/will be fixed with C++17 and inline static constexpr variables

**Comment by masterleinad** [comment]
> IMO, all the casting to bool for things like enum : bool { accessible = ... }; are bad changes. They can/will be fixed with C++17 and inline static constexpr variables

i agree but a follow-up pull request searching for `static_cast<bool>` is easy enough. I'm also fine with fixing here after the release when requiring C++17.

**Comment by dalg24** [comment]
I'd prefer the latter

**Comment by masterleinad** [comment]
> I'd prefer the latter

I'm not surprised. :slightly_smiling_face:

**Comment by masterleinad** [comment]
@dalg24 I simply replaced the relevant unnamed enums with `static constexpr bool` in [`4eafd28` (#4648)](https://github.com/kokkos/kokkos/pull/4648/commits/4eafd28ef5f1a0dc3bf590c38afd504e556f5dad). Even without requiring C++17 that should be fine as long as they are not ODR-used (which is impossible with the enums).

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

File: core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp [hunk 15]
```diff
@@ -501,7 +501,7 @@ struct CudaReductionsFunctor<FunctorType, ArgTag, false, true> {
           1;
     }
     bool is_last_block = false;
-    if (__syncthreads_or(num_teams_done == gridDim.x)) {
+    if (__syncthreads_or(num_teams_done == gridDim.x ? 1 : 0) != 0) {
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp:504]: if you must cast to `int`
```suggestion
    if (__syncthreads_or(static_cast<int>(num_teams_done == gridDim.x)) != 0) {
```

Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp:504]: I don't care about that too much (even if I prefer the current version for clarity).

**Comment by dalg24** [comment]
This one is wildly out of date and not being worked on currently.  Closing.

**Pull request closed without merging**